executor/executor.go

```diff
@@ -765,10 +765,23 @@ func (e *ShowDDLJobQueriesWithRangeExec) Next(ctx context.Context, req *chunk.Ch
 		return nil
 	}
 	numCurBatch := mathutil.Min(req.Capacity(), len(e.jobs)-e.cursor)
-	for i := e.cursor; i < e.cursor+numCurBatch; i++ {
+	appendedJobID := make(map[int64]struct{})
+	i := e.cursor
+	for {
 		if i >= int(e.offset) && i < int(e.offset+e.limit) {
-			req.AppendString(0, strconv.FormatInt(e.jobs[i].ID, 10))
-			req.AppendString(1, e.jobs[i].Query)
+			// check whether the job has been read
+			// for situations when this operation happens at the same time with new DDLs being executed
+			if _, ok := appendedJobID[e.jobs[i].ID]; ok {
+				numCurBatch++
```

Comment:
I prefer this variable not be incremented after assignment. It means in this `Next` invocation how many rows should be written. We can use another variable to count how many rows we have written.

Or we can filter the duplicated job at `Open`

Also seems ShowDDLJobQueriesExec also have this problem, please also fix it

Reply:
Here `numCurBatch` indeed means in this `Next` invocation how many rows should be written, instead of the number of rows we have written.
Line 774 is the situation where the newly retrieved job ID has already been in the key of map `appendedJobID`, which means we have retrieved a duplicate history job (due to the execution of new DDLs at the same time), so the handling method is to ignore this current line and increase `numCurBatch` by 1 to retrieve one more line in order to make the length of `req` conform to the number of rows that should be written. `i` is the variable that serve as the local cursor and keep count of how many rows we have written.
And in `Next` of `ShowDDLJobQueriesExec` there seems to be no increasing `numCurBatch`.

Reply:
> so the handling method is to ignore this current line and increase `numCurBatch` by 1

This breaks the definition. If we skip writing a row, "how many rows should be written" is not changed, but in current code we act like we did write the row and increase `numCurBatch`. In other words, take the ending condition of loop as `written == numCurBatch`, when a row is skipped we should neither increase `written` and `numCurBatch`. 

Another way to code is decrease `numCurBatch` by 1 after written a row, and the loop ends when `numCurBatch == 0`

`ShowDDLJobQueriesExec` may also have duplicated job in `e.jobs`, I think it may output a row twice.

Reply:
Modified. Here I use a variable `numOfRowsToWrite` to keep track of how many rows that we still need to write.